File: library/src/main/java/io/github/anderscheow/validator/rules/common/LengthRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.util.Locale;

import io.github.anderscheow.validator.rules.BaseRule;

public class LengthRule extends BaseRule {

    private int minLength;
    private int maxLength;

    public LengthRule(int minLength, int maxLength) {
        super(String.format(Locale.getDefault(), "Length must be between %d and %d", minLength, maxLength));
        this.minLength = minLength;
        this.maxLength = maxLength;
    }

    public LengthRule(int minLength, int maxLength, @StringRes int errorRes) {
        super(errorRes);
        this.minLength = minLength;
        this.maxLength = maxLength;
    }

    public LengthRule(int minLength, int maxLength, @NonNull String errorMessage) {
        super(errorMessage);
        this.minLength = minLength;
        this.maxLength = maxLength;
    }

    @Override
    public boolean validate(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        if (value instanceof String) {
            assertMinMax(minLength, maxLength);

            int length = ((String) value).length();

            boolean isMinValid = true;
            if (minLength != Integer.MIN_VALUE) {
                isMinValid = length >= minLength;
            }

            boolean isMaxValid = true;
            if (maxLength != Integer.MAX_VALUE) {
                isMaxValid = length <= maxLength;
            }

            return isMinValid && isMaxValid;
        }

        throw new ClassCastException("Required String value");
    }

    private void assertMinMax(int min, int max) {
        if (min > max) {
            String message = String.format(Locale.getDefault(), "'minLength' (%d) " +
                    "should be smaller than 'maxLength' (%d)", min, max);
            throw new IllegalStateException(message);
        }
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/LengthRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule
import java.util.*

class LengthRule : Rule {

    private var minLength: Int
    private var maxLength: Int

    constructor(minLength: Int, maxLength: Int, @StringRes errorRes: Int) :
            super(errorRes) {
        this.minLength = minLength
        this.maxLength = maxLength
    }

    constructor(minLength: Int, maxLength: Int, errorMessage: String) :
            super(errorMessage) {
        this.minLength = minLength
        this.maxLength = maxLength
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            value.let {
                assertMinMax(minLength, maxLength)

                val length = it.length

                var isMinValid = true
                if (minLength != Integer.MIN_VALUE) {
                    isMinValid = length >= minLength
                }

                var isMaxValid = true
                if (maxLength != Integer.MAX_VALUE) {
                    isMaxValid = length <= maxLength
                }

                return isMinValid && isMaxValid
            }
        }
    }

    private fun assertMinMax(min: Int, max: Int) {
        if (min > max) {
            val message = String.format(
                Locale.getDefault(),
                "'minLength' (%d) " + "should be smaller than 'maxLength' (%d)",
                min,
                max
            )
            throw IllegalStateException(message)
        }
    }
}

fun withinRange(minLength: Int, maxLength: Int, @StringRes errorRes: Int): LengthRule =
    LengthRule(minLength, maxLength, errorRes)

fun withinRange(minLength: Int, maxLength: Int, errorMessage: String): LengthRule =
    LengthRule(minLength, maxLength, errorMessage)
